### PR TITLE
Only list runtime dependencies when doing `npm ls --production`

### LINKED ISSUE #11559: `npm ls --production` contains dev dependencies
When I execute `npm ls --prod`, the output tree contains dev dependencies. My npm version is: **3.7.2** (npm below 3.7 is OK).

Example package.json:

```
...
  "devDependencies": {
    "mocha": "~2.4.5"
  },
  "dependencies": {
    "find-process": "~1.0.3"
  }
...
```

Result of `npm ls --prod --depth 1`

```
$ npm ls --prod --depth 0
demo@1.0.0 /Users/zoujie/demo
├── find-process@1.0.3
└── mocha@2.4.5 extraneous

npm ERR! extraneous: mocha@2.4.5 /Users/zoujie/demo/node_modules/mocha
```

When `--prod(uction)` arg was assign, the dev dependency mocha shouldn't be listed in the result.

After looked up the commit history, I find out the production filter logic has been removed: https://github.com/npm/npm/commit/9ab8b8d047792612ae7f9a6079745d51d5283a53. It fixed #10820, but missed this scenario.

### PR DESCRIPTION
**Pull request by yibn2008**
fix #11559

### CODE REVIEW

**Comment by othiym23** [comment]
@iarna, can you review this? This fixes more than just #11559, and has come up a few times recently.

File: lib/ls.js [hunk 1]
```diff
@@ -117,12 +117,18 @@ function filterByEnv (data) {
...
   Object.keys(data.dependencies).forEach(function (name) {
-    var keys = Object.keys(devDependencies)
-    if (dev && !production && keys.indexOf(name) === -1) return
-    if (!dev && keys.indexOf(name) !== -1 && data.dependencies[name].missing) return
-    dependencies[name] = data.dependencies[name]
+    if (!dev && devKeys.indexOf(name) !== -1 && data.dependencies[name].missing) {
+      return
+    }
+
+    if ((dev && devKeys.indexOf(name) >= 0) ||            // only --production
```
Comment (iarna) [inline lib/ls.js]: Am I misreading this or are the comments swapped between this line and the line following?

File: lib/ls.js [hunk 2]
```diff
@@ -117,12 +117,18 @@ function filterByEnv (data) {
...
   var dependencies = {}
-  var devDependencies = data.devDependencies || []
+  var devKeys = Object.keys(data.devDependencies || [])
+  var prodKeys = Object.keys(data._dependencies || [])
   Object.keys(data.dependencies).forEach(function (name) {
-    var keys = Object.keys(devDependencies)
-    if (dev && !production && keys.indexOf(name) === -1) return
-    if (!dev && keys.indexOf(name) !== -1 && data.dependencies[name].missing) return
-    dependencies[name] = data.dependencies[name]
+    if (!dev && devKeys.indexOf(name) !== -1 && data.dependencies[name].missing) {
```
Comment (iarna) [inline lib/ls.js]: This uses `!== -1` where as later you use `>= 0`– I'll standardize these when I merge this.

**Comment by iarna** [comment]
Looks solid to me 🐑

File: lib/ls.js (hunk 1 above)
Comment (yibn2008) [inline lib/ls.js]: Oh ... yes, the comments should be swapped, it's a mistake.

**Comment by zkat** [comment]
This was merged and included in `npm@3.7.4`.

**Pull request closed without merging**